File: lib/domain/chats/chats_repository.dart
import 'package:dartz/dartz.dart';
import 'package:telleo/domain/chats/message_entity.dart';
import 'package:telleo/domain/chats/message_packet.dart';
import 'chats_failures.dart';
import 'chat_entity.dart';

abstract class ChatsRepository {
  Future<Either<ChatsFailure, List<ChatEntity>>> getChats();
  Future<Either<ChatsFailure, ChatEntity>> createChat(String withId);
  Future<Either<ChatsFailure, ChatEntity>> updateChat(ChatEntity chat);
  Future<Either<ChatsFailure, Unit>> sendMessage(
      ChatEntity chat, String content);
  Stream<MessagePacket> onMessageReceived({String? chatId});
}

File: backend/src/api/chats/domain/chats_repository.ts
import { Either } from "funpjs";

import ChatEntity from "./chat_entity";
import ChatFailure from "./chat_failure";
import MessageEntity from "./message_entity";

export default interface ChatsRepository {
  getChats(uid: string): Promise<ChatEntity[]>;
  getChat(id: string): Promise<Either<ChatFailure, ChatEntity>>;
  exists(users: string[]): Promise<boolean>;
  createChat(users: string[]): Promise<Either<ChatFailure, ChatEntity>>;
  updateChat(chat: ChatEntity): Promise<Either<ChatFailure, ChatEntity>>;
  addMessage(
    chatId: string,
    message: MessageEntity
  ): Promise<Either<ChatFailure, ChatEntity>>;
}
